cli/rt/27_websocket.js

```diff
@@ -0,0 +1,290 @@
+        if (create.success) {
+          this.#rid = create.rid;
+          this.#extensions = create.extensions;
+          this.#protocol = create.protocol;
+          this.#readyState = this.#OPEN;
+          const event = new Event("open");
+          event.target = this;
+          this.onopen(event);
+          this.dispatchEvent(event);
+
+          this.#eventLoop();
+        } else {
+          this.#readyState = this.#CLOSED;
+
+          const errEvent = new Event("error");
+          errEvent.target = this;
+          this.onerror(errEvent);
+          this.dispatchEvent(errEvent);
+
+          const closeEvent = new CloseEvent("close");
+          closeEvent.target = this;
+          this.onclose(closeEvent);
+          this.dispatchEvent(closeEvent);
+        }
+      });
+    }
+
+    send(data) {
+      requiredArguments("WebSocket.send", arguments.length, 1);
+
+      if (
+        this.#readyState === this.#OPEN
+      ) {
+        if (data instanceof Blob) {
+          data.slice().arrayBuffer().then((buf) => {
+            this.#bufferedAmount += buf.byteLength;
+            sendSync("op_ws_send", {
+              rid: this.#rid,
+            }, buf);
+          });
+        } else if (
+          data instanceof Int8Array || data instanceof Int16Array ||
+          data instanceof Int32Array || data instanceof Uint8Array ||
+          data instanceof Uint16Array || data instanceof Uint32Array ||
+          data instanceof Uint8ClampedArray || data instanceof Float32Array ||
+          data instanceof Float64Array || data instanceof DataView
+        ) {
+          this.#bufferedAmount += data.byteLength;
+          sendSync("op_ws_send", {
+            rid: this.#rid,
+          }, data);
+        } else if (data instanceof ArrayBuffer) {
+          this.#bufferedAmount += data.byteLength;
+          sendSync("op_ws_send", {
+            rid: this.#rid,
+          }, new DataView(data));
+        } else {
+          const string = String(data);
+          const encoder = new TextEncoder();
+          this.#bufferedAmount += encoder.encode(string).byteLength;
+          sendSync("op_ws_send", {
+            rid: this.#rid,
+            text: string,
+          });
+        }
+      }
+    }
+
+    close(code, reason) {
+      if (code && (code !== 1000 && !(3000 <= code > 5000))) {
+        throw new DOMException(
+          "The close code must be either 1000 or in the range of 3000 to 4999.",
+          "NotSupportedError",
+        );
+      }
+
+      const encoder = new TextEncoder();
+      if (reason && encoder.encode(reason).byteLength > 123) {
+        throw new DOMException(
+          "The close reason may not be longer than 123 bytes.",
+          "SyntaxError",
+        );
+      }
+
+      if (this.#readyState === this.#CONNECTING) {
+        this.#readyState = this.#CLOSING;
+
+        sendAsync("op_ws_close", {
+          rid: this.#rid,
+          code,
+          reason,
+        }).then(() => {
```

Comment:
Why is it async op?

Reply:
Because the ws.close method is internally async (the method to invoke in js is sync, but there is a "close" event that gets fired in js once the close is successful)